iframe.html

```diff
@@ -13,39 +13,50 @@
 <body style="padding:0; margin:0; width:100%; height:100%; overflow:hidden">
   <div id='container' style='width:100%; height:100%;'></div>
   <script>
-    function queryStringToJSON(qs) {
-      var pairs = qs.split('&');
-      var result = {};
-      pairs.forEach(function(p) {
-        var pair = p.split('=');
-        var key = pair[0];
-        var value = decodeURIComponent(pair[1] || '');
-        if (value === "true") {
-          value = true;
-        } else if (value == "false") {
-          value = false;
-        } else if (!isNaN(parseFloat(value))) {
-          value = parseFloat(value);
-        } else if (value.match(/^\[.+\]$/)) {
-          value = JSON.parse(value);
-        } else if (value.match(/^\{.+\}$/)) {
-          value = JSON.parse(value);
-        } 
-
-        if (result[key]) {
-          if (Array.isArray(result[key])) {
-            result[key].push(value);
+    function queryStringToJSON(query) {
+      var setValue = function(root, path, value) {
+        if (path.length > 1){
+          var dir = path.shift();
+          if (typeof root[dir] == 'undefined') {
+              root[dir] = path[0] == '' ? [] : {};
+          }
+          arguments.callee(root[dir], path, value);
+        } else {
+          if (root instanceof Array) {
+            root.push(value);
           } else {
-            result[key] = [result[key], value];
+            root[path] = value;
           }
+        }
+      };
+
+      var nvp = query.split('&');
+      var data = {};
+      for (var i = 0 ; i < nvp.length ; i++) {
+        var pair = nvp[i].split('=');
+        var name = decodeURIComponent(pair[0]);
+        var value = decodeURIComponent(pair[1]);
+
+        var path = name.match(/(^[^\[]+)(\[.*\]$)?/);
+        var first = path[1];
```

Comment:
key-value: key is element #0 of pair, value is element #1, value may be encoded, so we want to decode it, so we deal with element #1 here, element #0 doesn't require decoding